Displaying Dogs in tableView via request+binding

simonrowlands opened the pull request:
Search bar isn't implemented yet
May add sections/subsections as dogs have sub-breeds

simonrowlands committed:
Pods: -RxDataSources +RxCocoa

simonrowlands committed:
+Dog Model +DogsTableView +APIRequest to fetch Dogs

simonrowlands committed:
Converting Dogs response into Dog array

simonrowlands committed:
Binding tableView to data

simonrowlands committed:
Added table/keyboard interactivity

tailec commented on NCU-Project/API/APIRequests.swift:
```diff
@@ -0,0 +1,23 @@
+import RxSwift
+import RxCocoa
+
+class APIRequests {
```
maybe call it `NetworkingAPI` or `DogsApi` or smth?

tailec replied:
also it could conform to some protocol so its easy to mock api in tests like `protocol Networking`

tailec commented on NCU-Project/API/APIRequests.swift:
```diff
@@ -0,0 +1,23 @@
+
+class APIRequests {
+    
+    static var shared = APIRequests()
```
no need for singleton, you can use normal initializer

tailec replied:
some people are afraid of singletons... :)

tailec commented on NCU-Project/API/APIRequests.swift:
```diff
@@ -0,0 +1,23 @@
+    
+    static var shared = APIRequests()
+    
+    let dogsAPIString = "https://dog.ceo/api/breeds/list/all"
```
maybe change name to `baseUrl` or `basePath`?

also it could be injected through initializer with default value like `init(basePath: String = "https://dog.ceo/api/")`
and rest of the path could be implemented in `getDogsJSON`

tailec replied:
for example we inject path in different projects because it may change on diffrent environments (staging, prod, uat etc.)

tailec commented on NCU-Project/API/APIRequests.swift:
```diff
@@ -0,0 +1,23 @@
+    
+    let dogsAPIString = "https://dog.ceo/api/breeds/list/all"
+    
+    func getDogsJSON() -> Observable<Any> {
```
`getDogs` looks better imo

tailec commented on NCU-Project/Dogs/DogsTableViewController.swift:
```diff
@@ -0,0 +1,45 @@
+    
+    private func bindViewModel() {
+        
+        tableView.dataSource = nil // Required as dataSource is being replaced
```
💯 for comment

tailec commented on NCU-Project/Dogs/DogsTableViewController.swift:
```diff
@@ -0,0 +1,45 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+//        let dogsJSON = viewModel.transform(DogsViewModel.Input(filterString: ""))
```
comment can go away

tailec commented on NCU-Project/Dogs/DogsViewModel.swift:
```diff
@@ -0,0 +1,44 @@
+            }
+            .map { anyDict -> [Dog] in
+                
+                guard let messageDict = anyDict as? [String : Any], let dict = messageDict["message"] as? [String : [String]] else {
```
can we use `Decodable` or something?

tailec replied:
also, you could move mapping somewhere else. I think `NetworkingApi` would be a good place for this now

tailec commented on NCU-Project/Model/Dog.swift:
```diff
@@ -0,0 +1,44 @@
+
+import Foundation
+
+class Dog {
```
make it `struct`, immutable by default!

tailec commented on NCU-Project/Model/Dog.swift:
```diff
@@ -0,0 +1,44 @@
+class Dog {
+    
+    var breed: String
+    var subBreeds: [String]
```
you can replace vars with `let`

tailec commented on NCU-Project/Dogs/DogsTableViewController.swift:
```diff
@@ -0,0 +1,45 @@
+import RxSwift
+import RxCocoa
+
+class DogsTableViewController: UITableViewController {
```
we can use `final` here

tailec replied:
oh didn't notice this before, I wouldn't recomment using `UITableViewController`. its more flexible to just create `UITableView`in case we want to add custom functioanlity into it

simonrowlands committed:
Integrating proposed changes:
Moving JSON logic into NetworkingAPI
TableViewController -> ViewController
Dog Class -> Struct

tailec commented on NCU-Project/Dogs/DogsViewController.swift:
```diff
@@ -0,0 +1,45 @@
+import RxSwift
+import RxCocoa
+
+final class DogsViewController: UIViewController, UIScrollViewDelegate {
```
why do you need `UIScrollViewDelegate` here? 🤔

tailec commented on NCU-Project/Main/MainTableViewController.swift:
```diff
@@ -0,0 +1,11 @@
+
+import UIKit
+
+class MainTableViewController: UITableViewController { }
```
looks empty, can we remove it?

tailec commented on NCU-Project/Model/Dog.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+struct Dog: Decodable {
```
`Decodable` not used

tailec commented on NCU-Project/API/NetworkingAPI.swift:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+    
+    static func getDogsJSON() -> Observable<[Dog]> {
```
its not JSON anymore :)

tailec reviewed (changes requested):
looks good now, just some questions

simonrowlands committed:
Removed unused viewController class
Removed unused decodable attribute
Removed broken table delegate and the controllers scrollDelegate conformation
Renamed some funcs

amlcurran reviewed (approved):
Looks good!

There is a couple of comments I _would_ have made, but they're things that will come up in the later modules of the NCU. Perhaps we could revisit this code again in a few modules time and there will be a couple of things to improve on with the knowledge from there!
